actors/market/src/state.rs

```diff
@@ -551,20 +572,163 @@ impl State {
+    ) -> Result<(), ActorError>
+    where
+        BS: Blockstore,
+    {
+        let mut provider_sectors = self.load_provider_sectors(store)?;
+        let mut sector_deals = load_provider_sector_deals(store, &provider_sectors, provider)?;
+
+        for (sector_number, deals) in sector_deal_ids {
+            let mut new_deals = deals.clone();
+            let existing_deal_ids = sector_deals
+                .get(sector_number)
+                .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to read sector deals")?;
+            if let Some(existing_deal_ids) = existing_deal_ids {
+                new_deals.deals.extend(existing_deal_ids.deals.iter());
+            }
+            new_deals.deals.sort();
+            new_deals.deals.dedup();
+            sector_deals
+                .set(sector_number, new_deals)
+                .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+                    format!("failed to set sector deals for {} {}", provider, sector_number)
+                })?;
+        }
+
+        save_provider_sector_deals(&mut provider_sectors, provider, &mut sector_deals)?;
+        self.save_provider_sectors(&mut provider_sectors)?;
+        Ok(())
+    }
+
+    // Reads and removes the sector deals mapping for an array of sector numbers,
+    pub fn pop_sector_deal_ids<BS>(
+        &mut self,
+        store: &BS,
+        provider: &Address,
+        sector_numbers: impl Iterator<Item = SectorNumber>,
+    ) -> Result<Vec<(SectorNumber, SectorDealIDs)>, ActorError>
+    where
+        BS: Blockstore,
+    {
+        let mut provider_sectors = self.load_provider_sectors(store)?;
+        let mut sector_deals = load_provider_sector_deals(store, &provider_sectors, provider)?;
+
+        let mut popped_sector_deals = Vec::new();
+        for sector_number in sector_numbers {
+            let deals: Option<SectorDealIDs> = sector_deals
+                .delete(&sector_number)
+                .with_context(|| format!("provider {}", provider))?;
+            if let Some(deals) = deals {
+                popped_sector_deals.push((sector_number, deals.clone()));
+            }
+        }
+
+        // Flush if any of the requested sectors were found.
+        if !popped_sector_deals.is_empty() {
+            if sector_deals.is_empty() {
+                // Remove from top-level map
+                provider_sectors
+                    .delete(provider)
+                    .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+                        format!("failed to delete sector deals for {}", provider)
+                    })?;
+            } else {
+                save_provider_sector_deals(&mut provider_sectors, provider, &mut sector_deals)?;
+            }
+            self.save_provider_sectors(&mut provider_sectors)?;
+        }
+
+        Ok(popped_sector_deals)
+    }
 
-        Ok(rval_pending_deal)
+    // Removes specified deals from the sector deals mapping.
+    // Missing deals are ignored.
```

Comment:
Do we expect missing deals?

Reply:
Probably not, but I think "ensure removed" is a reasonable semantic. This is called from cron as deals expire (soon to be removed from cron though). I'm not sure that failing here is any better than achieving the desired outcome if the prior state was inconsistent.